geps/gep-1619.md

```diff
@@ -163,43 +163,299 @@ Java application servers such as Tomcat and Jetty, were the first to standardize
+(clusters), thereby augmenting routing behavior in scenarios involving traffic splitting.
+
+#### Session Persistence API with GAMMA
+
+The object of the [GAMMA (Gateway API for Mesh Management and Administration)](https://gateway-api.sigs.k8s.io/contributing/gamma/)
+initiative is to provide support for service mesh and mesh-adjacent use-cases with Gateway API. Mesh traffic routing with
+Mesh traffic routing with Gateway API is accomplished with `xRoutes` specifying service-to-service traffic opposed to
+gateway-to-service traffic.
+
+Given that service meshes commonly have session persistence requirements, this API design should take into consideration
+session persistence needs in GAMMA and service mesh scenarios.
+
+#### API Design
+
+Since session persistence needs to augment a service, we implement as a metaresource policy as described in [Policy Attachment](https://gateway-api.sigs.k8s.io/references/policy-attachment/#direct-policy-attachment).
+The metaresource is named `SessionPersistencePolicy` and is only responsible for configuring session persistence for
+services. It is defined as a [Direct Policy Attachment](https://gateway-api.sigs.k8s.io/v1alpha2/references/policy-attachment/#direct-policy-attachment)
+without defaults or overrides, applied to the service it is targeting.
+
+**Cookie Attributes**
+
+A cookie consists of a collection of attributes. These attributes are pairs of key=value, where the value is optional
+for certain attributes. The cookie name attribute is the only mandatory attribute while all other cookie attributes
+are optional.
+
+The cookie attributes defined by [RFC6265](https://www.rfc-editor.org/rfc/rfc6265#section-5.2) are:
+- Name=_value_
 - Expires=_date_
 - Max-Age=_number_
 - Domain=_domain_
 - Path=_path-value_
 - Secure
 - HttpOnly
 
-However, there are de-facto standard attributes that have been proposed as updates to [rfc6265](https://www.rfc-editor.org/rfc/rfc6265#section-5.2):
-- SameSite=_[Strict|Lax|None]_ ([Draft RFC Update](https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-cookie-same-site-00))
-- Partitioned ([Draft RFC Update](https://www.ietf.org/archive/id/draft-cutler-httpbis-partitioned-cookies-01.html))
-
-**Session Cookie Strategy**
-
-Session cookie strategy is configuration specific to the creation of the cookie by the dataplane. These are not specific configuration parameters of the cookie itself, but instead the operation to take when inserting the cookie such as:
-- Rewriting an existing cookie value
-- Inserting a new cookie
-- Prefixing an existing cookie
-- Preserving an existing cookie value
-- Only adding cookies on specific HTTP methods (POST/GET)
+Other cookie attributes not defined by RFC6265, but are captured in draft RFCs and could be considered de facto
+standards due to wide acceptance are:
+- SameSite=[Strict|Lax|None]
+- Partitioned
+
+We implement core support for the `Name`, `Path`, and `TTL` attributes as the metrics in [implementations](#Implementations)
```

Comment:
I'm trying to understand your position to not expose Cookie configuration:

1. If App Developers aren't configuring the details/attributes of the Cookies, are you implying there is no other GWAPI Persona that would be able to configure cookies attributes? I.e. cookies attributes are not configurable in the entirety of GWAPI?
2. If it's an implementation detail and not exposed, then the implementations are deciding the cookie attributes (path, secure, httponly, samesite,...) to set/add/rewrite, correct?
3. How would a user of Gateway API change the cookie attributes then? Are you saying a user of Gateway API will never need to adjust the cookie attributes (minus TTL as you mentioned)?
4. I don't think I understand `using existing LB infrastructure - including well-established infra for Servlet persistent sessions.`.  How would it prevent using existing LB infrastructure? Could you elaborate more on that?

Reply:
1 - yes, it is possible that the cookie name and content are configured at platform level ( global load balancers that run outside of clusters and are controlled by SREs ). There are cases where the cluster admin may control the cookie name / content - but at a cluster level. The actual implementation of persistent sessions needs to integrate with global load balancers and infra - some may encrypt the cookies, or implement cross-service sessions. 
Servlet sessions are a good example, I think they were reviewed in the prior art.

2. Yes for most of it. 

3. They can't. Some implementations may provide some extra knobs - but I don't think it's something we can assume or standardize. 

4. If we expose a know allowing a user to set the cookie for a specific backend - an existing  global load balancer that uses some specific techniques (encrypted cookies, etc ) will no longer be usable. Or in Tomcat/other servlet implementations, where the persistent cookie name is set per server, not with backend granularity.
For example google load balancers or CloudRun ( see the prior art doc ) use specific cookie names for persistent sessions.

Reply:
There is one thing I would like to point out that cookie name needs to be known to client applications which have to copy the cookie value from the first response to subsequent requests. Unless they know the cookie name they cannot do it reliably - may be they don't want to indiscriminately copy all cookies. What that means is the cookie name needs to be either configurable (in this API) or an implementation should expose the cookie name to the users so users can code their client app accordingly. But the downside then is the code is tied to that particular implementation. Hence having a configurable cookie name is desirable, no?

Reply:
> There is one thing I would like to point out that cookie name needs to be known to client applications which have to copy the cookie value from the first response to subsequent requests

@sanjaypujare Valid point, but sounds like there is going to be trade off made here, I don't know if we can have it both ways. I'm betting @costinm will still advocate for opaque cookie names, but I don't want to speak for him. 

> If we expose a know allowing a user to set the cookie for a specific backend - an existing global load balancer that uses some specific techniques (encrypted cookies, etc ) will no longer be usable. 

@costinm There's still one thing I'm not totally understanding. If, given what I think you are suggesting with an API, a Gateway API user can turn on session persistence using HTTP cookies, but can't specify the name (i.e. opaque), how does that support or solve your use case with global load balancers? Isn't the implementation then just going to select an implementation-specific cookie name that would also break the global load balancer use case?

I think I'm still struggling to understand the architecture in the use case with existing infrastructure. Is Gateway API programming the existing infrastructure (i.e. global load balanacers)? Or are you saying the existing infrastructure is auxiliary to Gateway API, and that GWAPI should not prevent the infrastructure from using cookies?

Reply:
@gcs278 there are 2 different things: the cookie name and cookie value. My suggestion was making the cookie name configurable from this API so that client apps can use the name to copy the cookie value. If the cookie name is not configurable then it's not end of the world but I would prefer to make it configurable.

The cookie value needs to be opaque and ideally users should not (be able to) decode or manipulate the contents. So this spec should not dictate the cookie value contents.

Reply:
Yes, I am advocating to allowing/supporting opaque cookie names - with user-controlled names as an optional feature. 

Assuming you have an external frontend LB that handles persistent sessions ( with some encryption key, etc ) - the implementation of the gateway behind the LB will need to be aware of the internal protocol/naming/requirements. For example google LBs support persistent cookies - with some well defined names and values that app developers/backends can't customize. Servlet is another example - cookie name is shared config between the servlet engine and LBs ( with default name defined in the servlet spec).

Usually it's not Gateway API programming the infra - but infra determining how the gateway is implemented and works with the rest of the infra. In some cases it may be possible for a multi-tenant infra to use multiple names ( defined by each Gateway API ), in some cases it may not be.

Reply:
Ack. I understand value as being opaque, but still unclear on why making name configurable breaks global load balancers and existing infra (see questions above).

Reply:
That reminds me: there could be cases where a single request goes through say 2 LBs and each one using a cookie. Not allowing separate names for the 2 LBs's cookies will be problematic as one cookie will be overwritten. So it will be good to have (and require?) configurable cookie names

Reply:
Costin must of had a [comment](https://github.com/kubernetes-sigs/gateway-api/pull/2159#discussion_r1272781250) in review because it just showed up between mine and Sanjay's, just FYI if the conversation seems confusing.

@costinm sorry, I'm still a bit confused. [Here](https://github.com/kubernetes-sigs/gateway-api/pull/2159#discussion_r1260087846) you seem to imply we should remove `HttpCookie` configuration, but here you say `...with user-controlled names as an optional feature.`.

Is the issue here with the `Name` field being `Support: Core`? Or is the issue with it just existing at all? It's `+optional` currently. Does `as an optional feature` == `optional API field` here?